### PR TITLE
perf(topk_varlen, attn_scores): upstream sync — gvr_2 hint-free engines (TRT-LLM #18410) + paged-MQA relu epilogue (DKG !27837)

### PR DESCRIPTION
**Pull request by dhiraj113**
## 📌 Description

Upstream sync of two kernel families FlashInfer vendors from NVIDIA sources, with every change measured same-node against the current tree before adoption (and two upstream heuristics rejected on measurement).

**1. Paged-MQA logits (`attn_scores`, fp8 + fp4 decode kernels) — DKG MR !27837 + d9adb3cfa77**
`relu(x) = (x + |x|) / 2` in the fp32 epilogue: two packed `FADD2` replace four scalar `max`. Bit-exact with the previous output. fp8 folds the /2 into the existing scale multiply; fp4 halves the register-resident fp32 weights once per q stage (upstream's per-output multiply form was 6-9% *slower* on B200; a four-tree bisect isolated it, see commit message). `tmem.free` receives the static column count.

**2. GVR V2 top-k (`top_k_varlen` backend `gvr_2`) — TRT-LLM #18410 `hint_free` + router follow-up**
- Hint-free calls run dedicated compiled engines (hint loads compiled out). The FlashInfer-local anchor table (`arange(k)` per (device, k), growth lock, keep-list, its share of `release_gvr2_resources`) is removed. Contract change: hint mode is a warm-up dimension for CUDA graphs (`warmup_varlen` compiles both; a hinted eager call no longer warms a hint-free capture — the capture raises the usual error).
- Per-family bracket policy, each choice measured (Rubin/B200, 3 interleaved reps, same-kernel control column): identity gather for main/clus and register VPT>=2, register fold only at VPT==1, DEG plans share one compiled object. Upstream's fold-for-every-VPT (2-14% slower on 43/90 cells) and compile-out semantics (4-6% slower) were not adopted.
- Router: the 4K<n<=8K register rung now also covers K=2048 up to 1024 rows — 15-39% faster than the slab on B200, 148-SM B300 **and** Rubin (upstream's SM103-only exception holds on SM100/SM107 too). The same extension for K<=1024 wins on mixed lengths but loses 5-26% on uniform full-length rows, so it was not taken.
- `auto`: `N <= top_k` (every row is the identity answer) prefers `radix_filter` (0.85 vs 1.1 us, both parts).
- Not adopted: `QC = QUADC` for every register plan (a wash within ~6% on Rubin and 148-SM B300), TRT-LLM #18702 prefill window mode (separate decision).

## 🔍 Related Issues

Follows #4986 / #4811 (gvr_2). DKG #61 (router rungs) gets a follow-up note with the K=2048 finding.

## 🚀 Pull Request Checklist

- [x] Pre-commit hooks pass.
- [x] Tests: `tests/topk_varlen` + `tests/attn_scores` on A100, L40S, H100, RTX 5080, DRIVE P2021, B100, computelab B200, computelab SM103 (150 SMs), Rubin GR100 — all green on the final tree (944 passed on B200/SM103; 491 on Rubin; 87-90 passed / rest skipped off-arch).
- [x] Perf: every number in the commit messages is a same-node interleaved A/B (min of 3 reps).

## Reviewer Notes

Kernel-side diff is const_expr-gated and mechanical against upstream; the review-worthy parts are the host `run_varlen` hint-free path, the removed table (and its tests), `route()`'s new rung line, and the two `_top_k_varlen_heuristic` lines.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added dedicated hint-free execution for variable-length top-k selection.
  - Expanded K=2048 routing support.
  - Warmup now prepares hinted and hint-free execution paths.

- **Performance**
  - Improved FP4 and FP8 paged MQA logits processing for FP32 outputs, including safer extreme-value handling.

- **Bug Fixes**
  - Improved handling of malformed hints, unavailable hint data, negative effective lengths, and resource cleanup.

- **Documentation**
  - Clarified sequence-length units, compression behavior, and valid-column calculations.

- **Tests**
  - Expanded routing, concurrency, resource, hint-free, and numerical boundary coverage.

### CODE REVIEW

**Comment by dhiraj113** [comment]
@flashinfer-bot run

File: flashinfer/attn_scores/kernels/fp4_paged_mqa_logits.py [hunk 1]
```diff
@@ -2513,17 +2529,28 @@ def kernel(
...
+                                        # FADD2 with a free |.| operand modifier
+                                        # replace four FMNMX (no packed f32x2 max
+                                        # exists). Exact for finite x, and the
+                                        # accumulator is a finite dot product of
+                                        # fp4 operands; the /2 lives in the fp32
+                                        # weights (halved at the q-stage preload; the
+                                        # FP4 SF is baked into the accumulator, so
+                                        # there is no scale multiply to fold it into).
+                                        # Ported from DKG MR !27837.
+                                        a0, a1 = cute.arch.add_packed_f32x2(
```
Comment (leejnau) [inline flashinfer/attn_scores/kernels/fp4_paged_mqa_logits.py]: This is not exact for every finite x: x + abs(x) can overflow before the folded /2 is applied. A valid MXFP4 dot can produce x = 2^127 using finite UE8M0 Q/K scales.  With weight 0.5, the previous path returns finite 2^126, while this addition produces 2^128 → inf; the cached 0.25 weight cannot recover it. The public API does not bound scale exponents. Could we retain the scalar max or use a range-safe formulation, and add an exponent-boundary regression test? The duplicate path at line 3040 needs the same fix.

File: flashinfer/attn_scores/kernels/fp8_paged_mqa_logits.py [hunk 2]
```diff
@@ -1834,17 +1834,26 @@ def kernel(
...
-                                            acc_vec[n0], cutlass.Float32(0.0)
+                                        # relu(x) = (x + |x|) / 2: two packed
+                                        # FADD2 with a free |.| operand modifier
+                                        # replace four FMNMX (no packed f32x2 max
+                                        # exists). Exact for finite x, and the
+                                        # accumulator is a finite dot product of
+                                        # fp8 operands; the /2 folds into the
+                                        # scale multiply at the store. Ported
+                                        # from DKG MR !27837.
+                                        a0, a1 = cute.arch.add_packed_f32x2(
```
Comment (leejnau) [inline flashinfer/attn_scores/kernels/fp8_paged_mqa_logits.py]: Moving the /2 to the final scale changes the finite overflow range because the weighted head accumulation now operates on 2 * relu(x). For example, with one dot score equal to 1, weight FLT_MAX, KV scale 0.5, and all other heads zero, the previous path returns finite FLT_MAX / 2. This path computes 2 * FLT_MAX → inf in the FMA, and the final 0.25 scale cannot recover it. The public API does not bound the weights. Could we apply the divisor before weighted accumulation or retain the previous scalar ReLU path, and add a boundary regression test? The other FP32 branches at lines 1914, 2281, and 2350 need the same fix.

File: flashinfer/topk_varlen/kernels/gvr2_topk_host.py [hunk 3]
```diff
@@ -322,11 +324,19 @@ def _reg(BLK, VPT, MINB, NBH):
...
+    #     this as an SM103-only exception; it holds on SM100 and SM107 too.
+    #     For K <= 1024 the same extension wins 20-40% on mixed lengths but
+    #     loses 5-26% on uniform full-length rows at n=4160 and n=8192, so
+    #     those keep the rule above.
     # Reported upstream for adoption as DKG issue #61.
     if n4 <= 2048:
         if wide:
             return _reg(1024, 2, 1, 2 * NB)
-        if b <= 2 * sms or (b <= 4 * sms and n4 >= 1536):
+        if b <= 2 * sms or (b <= 4 * sms and n4 >= 1536) or (k == 2048 and b <= 1024):
```
Comment (leejnau) [inline flashinfer/topk_varlen/kernels/gvr2_topk_host.py:339]: This expands the K=2048 register route through 1024 rows, but the committed tests only assert router selection; no GPU oracle test exercises the newly admitted 600–1024-row range. Could we add one hinted and one hint-free mixed-length correctness case around rows=600/1024, n=4160/8192, and k=2048? This would protect the new dispatch boundary against future kernel/router changes.

File: flashinfer/topk_varlen/topk_varlen.py [hunk 4]
```diff
@@ -1558,7 +1567,7 @@ def top_k_varlen(
         be a contiguous, 16-byte-aligned int32 CUDA tensor on
         ``logits.device`` of shape ``[seq_lens.shape[0], top_k]``: a hint
         that violates this is **discarded with a RuntimeWarning** and the
-        call runs hint-free (``gvr_2`` with its synthetic anchor; an explicit
+        call runs hint-free (``gvr_2`` with its hint-free engines; an explicit
         ``"gvr"`` request is refused).
     compress_ratio : int, optional
         KV-index compression factor (``1`` for DSv3.2, ``4`` for DSv4).
```
Comment (leejnau) [inline flashinfer/topk_varlen/topk_varlen.py]: Could we make the length units explicit here? When compress_ratio > 1, seq_lens must still contain total KV lengths in uncompressed token units; valid logit columns for row t are computed as (seq_len - next_n + t + 1) // compress_ratio. The current parameter documentation can be read as accepting lengths already expressed in compressed/logit-column units. For example, SGLang currently stores compressed_seq_lens = seq_lens // 4, so passing those with compress_ratio=4 would divide the search range twice. Please clarify this here and in the seq_lens parameter description.

**Review by kahyunnam — APPROVED** [review]

File: flashinfer/attn_scores/kernels/fp4_paged_mqa_logits.py (hunk 1 above)
Comment (dhiraj113) [reply flashinfer/attn_scores/kernels/fp4_paged_mqa_logits.py]: Agreed, the packed add is not range-safe for x >= 2^127. Fixed in d421d278: both paths now apply relu as one `max.NaN.f32` per head (exact and range-safe for every finite x, NaN-propagating like the scalar max, one FMNMX where `cutlass.max` lowered to compare + select), and the fp4 weight halving at the preload is removed (both sites, 2542 and 3040, plus the smem-path heads). Regression tests in test_attn_scores_adversarial.py: `test_adv_fp4_relu_exponent_boundary` (e2m1 6 x 4 with 2^58 UE8M0 scales -> the head-dim dot at 2^127.58; fails on the previous tree, passes now) and `test_adv_fp8_relu_exponent_boundary`. Perf (same-node B200, three trees): the range-safe form runs at the speed of the pre-#5288 compare + select epilogue (within noise on every cell); the fold's 1-4% gain (10% on the smallest fp4 cell) is given up. Rubin could not be re-measured today (both Rubin hosts are down, one awaiting an OS reboot after a GPU Xid 154); tables in the commit message.

File: flashinfer/attn_scores/kernels/fp8_paged_mqa_logits.py (hunk 2 above)
Comment (dhiraj113) [reply flashinfer/attn_scores/kernels/fp8_paged_mqa_logits.py]: Agreed on the weighted-accumulation range as well. Same commit d421d278: the `scale_val * 0.5` at the four fp8 store sites is gone, relu is `max.NaN.f32` before the weighted accumulation at all four fp32 sites (1845, 1914, 2281, 2350), so the accumulation runs on relu(x) with unscaled weights. Your example is the regression test `test_adv_fp8_relu_weight_boundary` (dot 1, weight FLT_MAX, KV scale 0.5 -> exactly FLT_MAX / 2; fails on the previous tree, passes now).

File: flashinfer/topk_varlen/kernels/gvr2_topk_host.py (hunk 3 above)
Comment (dhiraj113) [reply flashinfer/topk_varlen/kernels/gvr2_topk_host.py:339]: Added in 7c8154b1: `test_gvr2_k2048_register_rung_oracle`, rows in {600, 1024} x n in {4160, 8192} x {hinted, hint-free}, mixed lengths with the short-row edges (1 token, K, K + 1) and a full-width row, the route asserted to land on the (512, 4, 2) rung, results checked row by row against torch.topk (tie-aware) with the identity / -1 tail contract. Passes on B200, B300, B100 and DRIVE (Rubin hosts were down today; the route is SM-count independent for this rung).

File: flashinfer/topk_varlen/topk_varlen.py (hunk 4 above)
Comment (dhiraj113) [reply flashinfer/topk_varlen/topk_varlen.py]: Clarified in 8ff99cf2: `seq_lens` are documented as KV-token lengths regardless of `compress_ratio`, with the per-row column formula `(seq_lens[i] - next_n + t + 1) // compress_ratio` and the explicit warning that lengths already divided by the ratio (SGLang's compressed_seq_lens) would divide the search range twice; `compress_ratio` carries the same statement.

**Comment by dhiraj113** [comment]
@flashinfer-bot run

File: flashinfer/attn_scores/kernels/fp4_paged_mqa_logits.py (hunk 1 above)
Comment (dhiraj113) [reply flashinfer/attn_scores/kernels/fp4_paged_mqa_logits.py]: Rubin follow-up (the hosts came back): same three-way A/B on a Rubin GR100 (181 GB SKU), graph replay, min of 2 runs per tree, ms.

| kind | batch | ctx | next_n | pre-#5288 (compare+select) | fold (x+abs x)/2 | range-safe max.NaN | fold/range-safe | base/range-safe |
|---|---|---|---|---|---|---|---|---|
| fp4 | 32 | 16384 | 1 | 0.0141 | 0.0136 | 0.0141 | 0.965x | 1.000x |
| fp4 | 32 | 65536 | 2 | 0.0346 | 0.0301 | 0.0346 | 0.870x | 1.000x |
| fp4 | 128 | 65536 | 2 | 0.0944 | 0.0792 | 0.0944 | 0.839x | 1.000x |
| fp4 | 128 | 131072 | 2 | 0.1721 | 0.1435 | 0.1720 | 0.834x | 1.001x |
| fp4 | 32 | 131072 | 4 | 0.0769 | 0.0688 | 0.0768 | 0.896x | 1.001x |
| fp4 | 128 | 131072 | 4 | 0.2443 | 0.2279 | 0.2442 | 0.933x | 1.000x |
| fp8 | 128 | 131072 | 1 | 0.2299 | 0.2289 | 0.2300 | 0.995x | 1.000x |
| fp8 | 32 | 16384 | 4 | 0.0216 | 0.0211 | 0.0216 | 0.977x | 1.000x |
| fp8 | 128 | 65536 | 4 | 0.1760 | 0.1727 | 0.1759 | 0.982x | 1.001x |

On Rubin the fold was 2-17% faster than the range-safe form (fp4 the most), while the range-safe form again matches the pre-#5288 compare + select epilogue on every cell. So the fix costs the fold's gain on both parts and regresses nothing against main. If that gain is wanted back safely, the way is to halve the *input* before the packed add (fold 0.5 into the block scale / KV scale upstream of the accumulator; `(x/2 + |x/2|)` is exactly relu(x) and cannot overflow) rather than the output; I have noted that on the DKG side for the upstream kernel and left it out of this PR.

File: flashinfer/topk_varlen/topk_varlen.py (hunk 4 above)
Comment (leejnau) [reply flashinfer/topk_varlen/topk_varlen.py]: Thanks the length units are clear now. One small correction: padded/evicted slots can make the numerator negative, and the GVR2 kernels clamp those rows to zero. Could we document the valid-column count as max(0, (seq_lens[i] - next_n + t + 1) // compress_ratio) in both descriptions?

Comment (dhiraj113) [reply flashinfer/topk_varlen/topk_varlen.py]: Done in 350ca430: both descriptions now give the count as `max(0, (seq_lens[i] - next_n + t + 1) // compress_ratio)`, with a note that padded / evicted slots make the numerator negative and such rows come out as all -1.

**Comment by dhiraj113** [comment]
@flashinfer-bot run

File: flashinfer/topk_varlen/topk_varlen.py (hunk 4 above)
Comment (leejnau) [reply flashinfer/topk_varlen/topk_varlen.py]: Thanks this implements my suggestion exactly. On a broader backend pass, I realized my suggestion assumed all backends already lower-clamp, which is not true: radix and gvr can write from negative output offsets, while radix_cutlass can reinterpret the negative length as unsigned. Sorry I missed that initially. Could we either clamp those three paths and add a cross-backend seq_len < next_n test, or scope the empty-row statement to GVR2?

Comment (dhiraj113) [reply flashinfer/topk_varlen/topk_varlen.py]: Good catch, went with the first option in f10fe0ae: radix and gvr now clamp the per-row length at zero in the kernel (the identity epilogue's `-1` fill no longer starts at a negative offset), and `_run_radix_cutlass` clamps the host-side lengths with `clamp(min=0, max=N)` so the uint32 kernel path never receives a negative count. radix_filter's per-row trivial path already treated a non-positive length as an empty row, so it only gains coverage.

New `test_seq_len_below_next_n_all_backends` runs next_n=3 with requests of length 0, 1, 2 (numerators down to -2) across all five backends. The output sits in a sentinel-filled arena with guard rows and the zero-length request comes first, so row 0's stray writes land in the guard row where nothing repairs them (behind a full row they were racily overwritten by that row's own kernel, which is why the bug went unnoticed). Without the clamps the test fails on radix and gvr, both with two `-1`s written into the guard row; in my probe the unfixed radix_cutlass rows still came out all -1, so that clamp is defensive. With the clamps it passes on B100/DRIVE/B200/B300/RTX 5080 and Rubin, and the full topk_varlen + attn_scores suites are green on the same set. The docs statement now holds for every backend, so I left it unscoped.

**Comment by dhiraj113** [comment]
@flashinfer-bot run

**Review by leejnau — APPROVED** [review]
